src/Accounts/Authenticators/ManagedServiceIdentityAuthenticator.cs

```diff
@@ -0,0 +1,91 @@
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.Azure.Commands.Common.Authentication;
+using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
+
+namespace Microsoft.Azure.PowerShell.Authenticators
+{
+    public class ManagedServiceIdentityAuthenticator : DelegatingAuthenticator
+    {
+        public const string CommonAdTenant = "organizations",
+            AppServiceManagedIdentityFlag = "AppServiceManagedIdentityFlag",
+            DefaultMSILoginUri = "http://169.254.169.254/metadata/identity/oauth2/token",
+            DefaultBackupMSILoginUri = "http://localhost:50342/oauth2/token";
+
+        public override Task<IAccessToken> Authenticate(AuthenticationParameters parameters)
+        {
+            var msiParameters = parameters as ManagedServiceIdentityParameters;
+            return Task.Run(() => GetManagedServiceToken(msiParameters.Account,
+                                                         msiParameters.Environment,
+                                                         msiParameters.TenantId,
+                                                         msiParameters.ResourceEndpoint));
+        }
+
+        public override bool CanAuthenticate(AuthenticationParameters parameters)
+        {
+            return parameters is ManagedServiceIdentityParameters;
+        }
+
+        private IAccessToken GetManagedServiceToken(IAzureAccount account, IAzureEnvironment environment, string tenant, string resourceId)
+        {
+            if (environment == null)
+            {
+                throw new InvalidOperationException("Environment is required for MSI Login");
+            }
+
+            if (!account.IsPropertySet(AzureAccount.Property.MSILoginUri))
+            {
+                account.SetProperty(AzureAccount.Property.MSILoginUri, DefaultMSILoginUri);
+            }
+
+            if (!account.IsPropertySet(AzureAccount.Property.MSILoginUriBackup))
+            {
+                account.SetProperty(AzureAccount.Property.MSILoginUriBackup, DefaultBackupMSILoginUri);
+            }
+
+            if (string.IsNullOrWhiteSpace(tenant))
+            {
+                tenant = environment.AdTenant ?? CommonAdTenant;
+            }
+
+            if (account.IsPropertySet(AppServiceManagedIdentityFlag))
+            {
+                return new ManagedServiceAppServiceAccessToken(account, environment, GetFunctionsResourceId(resourceId, environment), tenant);
```

Comment:
Would it be cleaner to have one authenticator for AppServiceMSI first in the chain, and the authenticator for Compute MSI after?